File: android/src/main/kotlin/com/appsprint/flutter/AppSprintFlutterPlugin.kt
package com.appsprint.flutter

import android.content.Context
import com.appsprint.sdk.AppSprint
import com.appsprint.sdk.AppSprintConfig
import com.appsprint.sdk.AppSprintEventType
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlin.concurrent.thread

class AppSprintFlutterPlugin : FlutterPlugin, MethodChannel.MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var context: Context

    private fun sdk(): AppSprint = AppSprint.shared(context)

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, "appsprint_flutter/native")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {

            // Core SDK

            "configure" -> {
                val apiKey = (call.argument<String>("apiKey") ?: "").trim()
                if (apiKey.isEmpty()) {
                    result.error("CONFIGURE_ERROR", "AppSprint.configure requires a non-empty apiKey.", null)
                    return
                }
                thread(start = true) {
                    try {
                        val config = AppSprintConfig(
                            apiKey = apiKey,
                            apiUrl = call.argument<String>("apiUrl") ?: "https://api.appsprint.app",
                            enableAppleAdsAttribution = call.argument<Boolean>("enableAppleAdsAttribution") ?: true,
                            isDebug = call.argument<Boolean>("isDebug") ?: false,
                            logLevel = call.argument<Int>("logLevel") ?: if (call.argument<Boolean>("isDebug") == true) 0 else 2,
                            customerUserId = call.argument<String>("customerUserId"),
                        )
                        sdk().configure(config)
                        result.success(null)
                    } catch (e: Exception) {
                        result.error("CONFIGURE_ERROR", e.message, null)
                    }
                }
            }

            "sendEvent" -> {
                val eventTypeStr = call.argument<String>("eventType") ?: "custom"
                thread(start = true) {
                    try {
                        val type = AppSprintEventType.entries.find { it.wireValue == eventTypeStr } ?: AppSprintEventType.CUSTOM
                        val name = call.argument<String>("name")
                        val params = mutableMapOf<String, Any?>()
                        call.argument<Map<String, Any?>>("parameters")?.forEach { (k, v) -> params[k] = v }
                        val revenue = call.argument<Double>("revenue")
                        val currency = call.argument<String>("currency")
                        if (revenue != null && revenue != 0.0) params["revenue"] = revenue
                        if (currency != null) params["currency"] = currency
                        sdk().sendEvent(type, name, if (params.isNotEmpty()) params else null)
                        result.success(null)
                    } catch (e: Exception) {
                        result.error("SEND_EVENT_ERROR", e.message, null)
                    }
                }
            }

            "sendTestEvent" -> {
                thread(start = true) {
                    try {
                        val r = sdk().sendTestEvent()
                        result.success(mapOf("success" to r.success, "message" to r.message))
                    } catch (e: Exception) {
                        result.error("TEST_EVENT_ERROR", e.message, null)
                    }
                }
            }

            "flush" -> {
                thread(start = true) {
                    try {
                        sdk().flush()
                        result.success(null)
                    } catch (e: Exception) {
                        result.error("FLUSH_ERROR", e.message, null)
                    }
                }
            }

            "clearData" -> {
                try {
                    sdk().clearData()
                    result.success(null)
                } catch (e: Exception) {
                    result.error("CLEAR_DATA_ERROR", e.message, null)
                }
            }

            "setCustomerUserId" -> {
                val userId = call.argument<String>("userId") ?: ""
                thread(start = true) {
                    try {
                        sdk().setCustomerUserId(userId)
                        result.success(null)
                    } catch (e: Exception) {
                        result.error("SET_USER_ID_ERROR", e.message, null)
                    }
                }
            }

            "enableAppleAdsAttribution" -> {
                sdk().enableAppleAdsAttribution()
                result.success(null)
            }

            "getAppSprintId" -> result.success(sdk().getAppSprintId())

            "getAttribution" -> {
                val attr = sdk().getAttribution()
                if (attr == null) {
                    result.success(null)
                    return
                }
                val map = mutableMapOf<String, Any?>(
                    "source" to attr.source,
                    "confidence" to attr.confidence,
                )
                attr.campaignName?.let { map["campaignName"] = it }
                attr.utmSource?.let { map["utmSource"] = it }
                attr.utmMedium?.let { map["utmMedium"] = it }
                attr.utmCampaign?.let { map["utmCampaign"] = it }
                result.success(map)
            }

            "isInitialized" -> result.success(sdk().isInitialized())

            "isSdkDisabled" -> result.success(sdk().isSdkDisabled())

            "destroy" -> {
                sdk().destroy()
                result.success(null)
            }

            // Utility

            "getDeviceInfo" -> {
                result.success(mapOf(
                    "deviceModel" to android.os.Build.MODEL,
                    "screenWidth" to context.resources.displayMetrics.widthPixels,
                    "screenHeight" to context.resources.displayMetrics.heightPixels,
                    "locale" to java.util.Locale.getDefault().toLanguageTag(),
                    "timezone" to java.util.TimeZone.getDefault().id,
                    "osVersion" to android.os.Build.VERSION.RELEASE,
                ))
            }

            "getAdServicesToken" -> result.success(null) // iOS only

            "requestTrackingAuthorization" -> result.success(false) // iOS only

            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}

File: ios/Classes/AppSprintFlutterPlugin.swift
import Flutter
import UIKit
import AppSprintSDK

public class AppSprintFlutterPlugin: NSObject, FlutterPlugin {

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "appsprint_flutter/native", binaryMessenger: registrar.messenger())
    let instance = AppSprintFlutterPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {

    // MARK: - Core SDK

    case "configure":
      guard let args = call.arguments as? [String: Any],
            let apiKey = args["apiKey"] as? String, !apiKey.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
        result(FlutterError(code: "CONFIGURE_ERROR", message: "AppSprint.configure requires a non-empty apiKey.", details: nil))
        return
      }
      Task { @MainActor in
        let enableAppleAds = args["enableAppleAdsAttribution"] as? Bool ?? true
        let isDebug = args["isDebug"] as? Bool ?? false
        let logLevelRaw = args["logLevel"] as? Int
        let customerUserId = args["customerUserId"] as? String

        let logLevel: AppSprintLogLevel
        if let raw = logLevelRaw, let level = AppSprintLogLevel(rawValue: raw) {
          logLevel = level
        } else {
          logLevel = isDebug ? .debug : .warn
        }

        var sdkConfig = AppSprintConfig(
          apiKey: apiKey,
          enableAppleAdsAttribution: enableAppleAds,
          isDebug: isDebug,
          logLevel: logLevel,
          customerUserId: customerUserId
        )

        if let urlString = args["apiUrl"] as? String, let url = URL(string: urlString) {
          sdkConfig = AppSprintConfig(
            apiKey: apiKey,
            apiURL: url,
            enableAppleAdsAttribution: enableAppleAds,
            isDebug: isDebug,
            logLevel: logLevel,
            customerUserId: customerUserId
          )
        }

        await AppSprint.shared.configure(sdkConfig)
        result(nil)
      }

    case "sendEvent":
      guard let args = call.arguments as? [String: Any],
            let eventTypeStr = args["eventType"] as? String else {
        result(FlutterError(code: "SEND_EVENT_ERROR", message: "eventType is required", details: nil))
        return
      }
      Task { @MainActor in
        let type = AppSprintEventType(rawValue: eventTypeStr) ?? .custom
        let name = args["name"] as? String
        var params: [String: Any]? = args["parameters"] as? [String: Any]

        if let rev = args["revenue"] as? Double {
          if params == nil { params = [:] }
          params?["revenue"] = rev
        }
        if let cur = args["currency"] as? String {
          if params == nil { params = [:] }
          params?["currency"] = cur
        }

        await AppSprint.shared.sendEvent(type, name: name, params: params)
        result(nil)
      }

    case "sendTestEvent":
      Task { @MainActor in
        let r = await AppSprint.shared.sendTestEvent()
        result(["success": r.success, "message": r.message])
      }

    case "flush":
      Task { @MainActor in
        await AppSprint.shared.flush()
        result(nil)
      }

    case "clearData":
      Task { @MainActor in
        AppSprint.shared.clearData()
        result(nil)
      }

    case "setCustomerUserId":
      guard let args = call.arguments as? [String: Any],
            let userId = args["userId"] as? String else {
        result(FlutterError(code: "SET_USER_ID_ERROR", message: "userId is required", details: nil))
        return
      }
      Task { @MainActor in
        await AppSprint.shared.setCustomerUserId(userId)
        result(nil)
      }

    case "enableAppleAdsAttribution":
      Task { @MainActor in
        AppSprint.shared.enableAppleAdsAttribution()
        result(nil)
      }

    case "getAppSprintId":
      Task { @MainActor in
        result(AppSprint.shared.getAppSprintId())
      }

    case "getAttribution":
      Task { @MainActor in
        guard let attr = AppSprint.shared.getAttribution() else {
          result(nil)
          return
        }
        var dict: [String: Any] = [
          "source": attr.source,
          "confidence": attr.confidence,
        ]
        if let c = attr.campaignName { dict["campaignName"] = c }
        if let s = attr.utmSource { dict["utmSource"] = s }
        if let m = attr.utmMedium { dict["utmMedium"] = m }
        if let c = attr.utmCampaign { dict["utmCampaign"] = c }
        result(dict)
      }

    case "isInitialized":
      Task { @MainActor in
        result(AppSprint.shared.isInitialized)
      }

    case "isSdkDisabled":
      Task { @MainActor in
        result(AppSprint.shared.isSdkDisabled())
      }

    case "destroy":
      Task { @MainActor in
        AppSprint.shared.destroy()
        result(nil)
      }

    // MARK: - Utility

    case "getDeviceInfo":
      Task { @MainActor in
        let info = AppSprintNative.getDeviceInfo()
        var dict: [String: Any] = [:]
        if let m = info.deviceModel { dict["deviceModel"] = m }
        if let w = info.screenWidth { dict["screenWidth"] = w }
        if let h = info.screenHeight { dict["screenHeight"] = h }
        if let l = info.locale { dict["locale"] = l }
        if let t = info.timezone { dict["timezone"] = t }
        if let o = info.osVersion { dict["osVersion"] = o }
        if let v = info.idfv { dict["idfv"] = v }
        if let a = info.idfa { dict["idfa"] = a }
        if let tk = info.adServicesToken { dict["adServicesToken"] = tk }
        result(dict)
      }

    case "getAdServicesToken":
      let token = AppSprintNative.getAdServicesToken()
      result(token as Any)

    case "requestTrackingAuthorization":
      Task {
        let authorized = await AppSprintNative.requestTrackingAuthorization()
        result(authorized)
      }

    default:
      result(FlutterMethodNotImplemented)
    }
  }
}
